runtime/src/comm/gasnet/comm-gasnet.c

```diff
@@ -1302,85 +1562,28 @@ void  chpl_comm_execute_on(c_nodeid_t node, c_sublocid_t subloc,
-    // AMRequest. We'd have to pack them as 32-bit arguments though.
-    // Alternatively, we could make it a stack-local variable. That
-    // would be OK since AMRequestMedium:
-    //   * doesn't need its payload to be in the registered memory segment
-    //   * allows us to re-use the source memory once it returns
-    // and since nothing in info actually ends up in the task we
-    // start (arg does but can be copied as a pointer)
-    info = (fork_t*)chpl_mem_allocMany(1, info_size,
-                                       CHPL_RT_MD_COMM_FRK_SND_INFO, 0, 0);
-    info->caller = chpl_nodeID;
-    info->subloc = subloc;
-    info->ack = &done;
-    info->serial_state = chpl_task_getSerial();
-    info->fid = fid;
-    info->arg_size = arg_size;
-
-    init_done_obj(&done, 1);
-
-    if (passArg) {
-      if (arg_size)
-        chpl_memcpy(&(info->arg), arg, arg_size);
-      GASNET_Safe(gasnet_AMRequestMedium0(node, FORK, info, info_size));
-    } else {
-      chpl_memcpy(&(info->arg), &arg, sizeof(void*));
-      GASNET_Safe(gasnet_AMRequestMedium0(node, FORK_LARGE, info, info_size));
-    }
-
-    wait_done_obj(&done);
-    chpl_mem_free(info, 0, 0);
+    execute_on_common(node, subloc, fid, arg, arg_size,
+                     /*fast*/ false, /*blocking*/ true);
   }
 }
 
 void  chpl_comm_execute_on_nb(c_nodeid_t node, c_sublocid_t subloc,
-                        chpl_fn_int_t fid, void *arg, size_t arg_size) {
-  fork_t *info;
-  size_t  info_size;
-  int     passArg = (chpl_nodeID == node
-                     || sizeof(fork_t) + arg_size <= gasnet_AMMaxMedium());
+                        chpl_fn_int_t fid,
+                        chpl_comm_on_bundle_t *arg, size_t arg_size) {
 
-  void* argCopy = NULL;
+  if (chpl_nodeID == node) {
+    chpl_bool serial_state = chpl_task_getSerial();
 
-  if (passArg) {
-    info_size = sizeof(fork_t) + arg_size;
-  } else {
-    info_size = sizeof(fork_t) + sizeof(void*);
-  }
-  info = (fork_t*)chpl_mem_allocMany(info_size, sizeof(char), CHPL_RT_MD_COMM_FRK_SND_INFO, 0, 0);
-  info->caller = chpl_nodeID;
-  info->subloc = subloc;
-  info->ack = info; // pass address to free after get in large case
-  info->serial_state = chpl_task_getSerial();
-  info->fid = fid;
-  info->arg_size = arg_size;
-  if (passArg) {
-    if (arg_size)
-      chpl_memcpy(&(info->arg), arg, arg_size);
-  } else {
-    // If the arg bundle is too large to fit in fork_t (i.e. passArg == false), 
-    // Copy the args into auxiliary memory and pass a pointer to this instead.
-    argCopy = chpl_mem_allocMany(1, arg_size,
-                                 CHPL_RT_MD_COMM_FRK_SND_ARG, 0, 0);
-    chpl_memcpy(argCopy, arg, arg_size);
-    *(void**)(&(info->arg)) = argCopy;
-  }
+    assert(0); // locale model code should prevent this...
 
-  if (chpl_nodeID == node) {
-    if (info->serial_state)
-      fork_nb_wrapper(info);
+    // Visual Debug?  Should we generate a task here???
```

Comment:
Won't we do the appropriate vdebug task callback in the tasking layer, if we make a task?  If we don't make a task here (i.e. if we do an ftable_call), we definitely should not do a vdebug task callback, that's certain.